google/cloud/storage/download_options.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_DOWNLOAD_OPTIONS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_DOWNLOAD_OPTIONS_H_
+
+#include "google/cloud/storage/internal/complex_option.h"
+#include <iostream>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+struct ReadRangeData {
+  std::int64_t begin;
+  std::int64_t end;
+};
+
+/**
+ * Request only a portion of the GCS object in a ReadObject operation.
+ *
+ * Note that the range is right-open, as it is customary in C++. That is, it
+ * excludes the `end` byte.
+ */
+struct ReadRange : public internal::ComplexOption<ReadRange, ReadRangeData> {
+  ReadRange() : ComplexOption() {}
+  explicit ReadRange(std::int64_t begin, std::int64_t end)
+      : ComplexOption(ReadRangeData{begin, end}) {}
```

Comment:
We decided long ago to avoid duplicating checks that the server is going to do anyway:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/master/doc/adr/2018-06-19-do-not-duplicate-server-side-validation.md

Reply:
I should have added: of course we can change our minds about these decisions, but when we do, we need to review them in a lot of places, that should be a separate PR and discussion.